wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerAbstractFileWatcher.java

```diff
@@ -0,0 +1,108 @@
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.languageserver;
+
+import com.google.common.collect.ImmutableList;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.languageserver.RegistryContainer.Registry;
+import org.eclipse.lsp4j.DidChangeWatchedFilesParams;
+import org.eclipse.lsp4j.FileChangeType;
+import org.eclipse.lsp4j.FileEvent;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.eclipse.lsp4j.services.WorkspaceService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Notifies registered language servers about file related events. Specifying of file event type is
+ * expected to be defined through this class inheritance.
+ *
+ * @author Dmytro Kulieshov
+ * @see <a
+ *     href="https://microsoft.github.io/language-server-protocol/specification#workspace_didChangeWatchedFiles">DidChangeWatchedFiles
+ *     Notification</a> section of LSP specification
+ */
+class LanguageServerAbstractFileWatcher implements Consumer<Path> {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(LanguageServerAbstractFileWatcher.class);
+
+  private final EventService eventService;
+  private final Registry<Set<PathMatcher>> pathMatcherRegistry;
+  private final FileChangeType fileChangeType;
+
+  private final Map<PathMatcher, Consumer<Path>> eventConsumers = new ConcurrentHashMap<>();
+
+  @Inject
+  LanguageServerAbstractFileWatcher(
+      EventService eventService, RegistryContainer registries, FileChangeType fileChangeType) {
+    this.eventService = eventService;
+    this.pathMatcherRegistry = registries.pathMatcherRegistry;
+    this.fileChangeType = fileChangeType;
+  }
+
+  @PostConstruct
+  protected void subscribe() {
+    eventService.subscribe(
+        event -> {
+          String id = event.getId();
+          Set<PathMatcher> pathMatchers = pathMatcherRegistry.getOrNull(id);
+          registerConsumers(event.getLanguageServer(), pathMatchers);
+        },
+        LanguageServerInitializedEvent.class);
+  }
+
+  @PreDestroy
+  void removeAllWatchers() {
+    eventConsumers.clear();
+  }
+
+  private void registerConsumers(LanguageServer languageServer, Set<PathMatcher> pathMatchers) {
+    if (pathMatchers == null) {
+      return;
+    }
+
+    for (PathMatcher pathMatcher : pathMatchers) {
```

Comment:
If you could look a few lines earlier you would see that `pathMatchers` is a `Set` so in general there can't be a situation when we have two "equal" path matchers here. 

I'm guessing you are talking more of a problem of that part of the [`RegistryContainer`](https://github.com/eclipse/che/blob/07263f1e30089689d71b057f747a44a29283e3c4/wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/RegistryContainer.java#L44) that stores path matchers, I tend to agree with you saying that we probably need a different data structure here.

As the remark is not related directly to this pull request I will skip it here. However please file an issue regarding erroneous path matcher registry if you think it is required.

Reply:
No, there are distinct sets of path matchers for different language servers which may interfere with each other.

Reply:
Oh, I think now I understood what you mean, however I'm not sure that the real situation is what you describe it, please note the line 
```java
Set<PathMatcher> pathMatchers = pathMatcherRegistry.getOrNull(event.getId());
```
This means that each set that we iterate in the line
```java
for (PathMatcher pathMatcher : pathMatchers) {
```
is for a single language server with id `event.getId()` and does not interfere with any other language server path matcher set.

Reply:
And when the next language server is initialized? The "eventConsumers" map is shared among _all_ language servers. And yes, different language servers will profess interest in the same files.

Reply:
You need either a Map<PathMatcher, List<EventConsumer>> or a List of Pair<PathMatcher, EventConsumer>. Since we never use a get() on the map anyway, there is also no performance advantage.

Reply:
Okay, now I see, thanks.